### DISCUSSION TITLE
Any change to describe a default source for a property that sets a new value on an update?

### DISCUSSION THREAD

**Discussion by elzorrorebelde** [geldata/gel#3180] [Q&A] [answered]
not sure how to ask this since I'm not a specialist on database things.

I have the following:

```
module default {
     abstract type DatedType {
          required property created_at -> datetime {
               default := std::datetime_current();
               readonly := true;
          };
          required property updated_at -> datetime {
               default := std::datetime_current();
          };
          property deleted_at -> datetime;
     }

     type User extending DatedType {
          required property username -> str {
               constraint exclusive;
          };
     }

     type Name extending DatedType {
          required property lang -> str;
          required property value -> str;

          constraint exclusive on ( (.lang, .value) );
     }
}
```

as you can see, I defined a type with fields that represent when the thing was created, when it was updated and when it was (soft) deleted. Setting a default upon  creation is kind of easy since we have `default` in there. If we had something like `default on update := std::datetime_current();` I could probably define `updated_at` as `readonly := true` as I did with `created_at` and let edgedb update that property for me when I'm updating some record.

Is this currently possible in some other way other than manually setting the new value on updates?

**Comment by elprans** [comment, marked as answer]
Currently, there isn't a schema-level way of implementing `updated_at` like you've described, so you'll need to keep including `updated_at := std::datetime_current()` in your `UPDATE` statements.

We are working on a general query rewrite system (see [roadmap](https://www.edgedb.com/roadmap#access-control)), which would allow arbitrary injection of operations into queries based on a condition (or unconditionally).  Here's how (tentatively) your schema would look:

```
module default {
     abstract type DatedType {
          required property updated_at -> datetime;
          property deleted_at -> datetime;

          rule bump_updated_at on update {
              update DatedType set { updated_at := datetime_current() }
          }
     }
}
```

Or, in a more general audit log-style system:

```
module default {
     abstract type DatedType {
          required property updated_at -> datetime;
          property deleted_at -> datetime;

          rule audit on update {
              insert AuditLog { text := "updated " ++ DatedType.__type__.name ++ " on " ++ datetime_current() }
          }
     }
}
```

We expect this (or some part of it) to land in 2.0.

**Reply by ifiokjr** [reply]
Thanks for the answer. Is this still on the roadmap?

**Reply by elprans** [reply]
Yes, though not for 2.0.

**Comment by nrathaus** [comment]
Was this addressed? we are at version 4.0? or are you referring to something else when you mention `2.0`

**Reply by raddevon** [reply]
[Mutation rewrites](https://www.edgedb.com/docs/datamodel/mutation_rewrites) were implemented in 3.0.

**Comment by nrathaus** [comment]
Thank you for the update, it wasn't clear from the issue above that it was - maybe it is a good idea to mark it at the title as [solved]?
